---
title: Usage
---

This page gives a detailed description of each component/subcommand of Tranquillyzer. For a quick overview of how to run
Tranquillyzer, see the [Quick Start](webpages/quick_start.qmd) guide.

# Input / Output

## Input

The raw input to Tranquillyzer is an absolute or relative path to the directory containing your raw FASTA or FASTQ files
for your data. The expected file extensions are `.fasta/.fa/.fasta.gz/.fa.gz` for FASTA files or
`.fastq/.fq/.fastq.gz/.fq.gz` for FASTQ files.

## Output

Tranquillyzer produces a variety of output files from its various subcommands. A few outputs to be particularly aware of
are:

- Demultiplexed FASTA files from `annotate-reads`
- `.parquet` files with valid and invalid annotations from `annotate-reads`
- `.pdf` files containing QC plots from `readlengthdist` and `visualize`
- A coordinate-sorted BAM from `align`
- A deduplicated, coordinate-sorted BAM from `dedup`

# Preprocessing

## Overview

To enhance the efficiency of the annotation process, Tranquillyzer organizes raw reads into separate `.parquet` files,
grouped based on their lengths. This approach optimizes data compression within each bin, accelerates the annotation of
the entire dataset, and facilitates the visualization of user-specified annotated reads without dependence on the
annotation status of the complete dataset. Tranquillyzer parallelizes the preprocessing step by distributing each input
file to its own CPU thread. Therefore, to maximize the benefits of parallelization, it is best to provide many small
input files rather than one large input file. Typically, this is how ONT provides the basecalled FASTA/FASTQ files from
its sequencing runs, so it is highly suggested to leave this structure as is.

## Usage

```bash
tranquillyzer preprocess \
    [OPTIONS]
    FASTA_DIR \
    OUTPUT_DIR
```

## Command Line Arguments

Required Arguments

- `FASTA_DIR`: Path to your RAW DATA directory
- `OUTPUT_DIR`: Path to your OUTPUT directory

Optional Arguments

- `--output-base-qual` / `--no-output-base-qual`: Whether to output base quality scores from the FASTQ file (default:
  `--no-output-base-qual`)
- `--chunk-size INTEGER`: Starting chunk size for processing, dynamically adjusted based on increasing read length
  (default: 100000)
- `--threads INTEGER`: Number of CPU threads (default: 12)
- `--help`: Print help message and exit

# Read Length Distribution

## Overview

As an initial quality control metric, users may wish to visualize the read length distribution. The `readlengthdist`
subcommand generates a plot with log<sub>10</sub>-transformed read lengths on the x-axis and their corresponding
frequencies on the y-axis. The output is provided as a `.png` file in the `plots/` subdirectory of the provided
`OUTPUT_DIR` path.

## Usage

```bash
tranquillyzer readlengthdist \
    [OPTIONS]
    OUTPUT_DIR
```

## Command Line Arguments

Required Arguments

- `OUTPUT_DIR`: Path to your OUTPUT directory

Optional Arguments

- `--help`: Print help message and exit

# Available Models

## Overview

Tranquillyzer provides a subcommand for viewing the available models (found within the `models/` directory) and the
corresponding read architecture and exact sequences (e.g., adapters, primers, etc.) used to train the model.

## Usage

```bash
tranquillyzer availablemodels \
    [OPTIONS]
```

## Command Line Arguments

Required Arguments

- None

Optional Arguments

- `--help`: Print help message and exit

# Annotation, Barcode Correction, and Demultiplexing

## Overview

The `annotate-reads` subcommand annotates the reads, extracts barcode sequences, corrects the barcodes, and assigns
reads to their respective cells (i.e., demultiplexes) in a single step. It produces the following outputs:

- Demultiplexed FASTA files located at `OUTPUT_DIR/demuxed_fasta/`
- Annotation metadata:
  - Valid reads can be found at `OUTPUT_DIR/annotations_valid.parquet`
  - Invalid reads can be found at `OUTPUT_DIR/annotations_invalid.parquet`
- Quality control (QC) plots
  - `OUTPUT_DIR/plots/barcode_plots.pdf`
  - `OUTPUT_DIR/plots/demux_plots.pdf`
  - `OUTPUT_DIR/plots/full_read_annots.pdf`

**Note:** Before running the `annotate-reads` subcommand, ensure you select the appropriate model and model type for
your dataset. Tranquillyzer supports multiple model types:

- REG (the base model, a standard CNN-LSTM model)
- CRF (CNN-LSTM model with an added CRF layer for improved label consistency)
- HYB (hybrid mode which runs REG first and reprocesses invalid reads with CRF)

The conventions for naming models are as follows:

- The REG model uses the base name (e.g., `10x3p_sc_ont_011.h5`)
- The CRF model (and the HYB model when it runs the CRF model) is the base name followed by `_w_CRF` (e.g.
  `10x3p_sc_ont_011_w_CRF.h5`)

To select a model type (REG, CRF, or HYB), simply specify the base model name. Tranquillyzer will automatically detect
the presence of the corresponding `_w_CRF` model file for the CRF or HYB types.

If only one version (REG or CRF) is available for a model, the user must select the corresponding model type explicitly.
We recommend verifying which model versions are present (via `availablemodels`) before running `annotate-reads`.

Currently available trained models can be downloaded from [this  Dropbox link](https://www.dropbox.com/scl/fo/3lms8n97bnufzqa4ausv9/AGkO3EVrL1ZgctwEwTK1mEA?rlkey=47m69a6smwsisdznbwu3jvjpu&st=253al7s3&dl=0)
and should be placed in the `models/` directory within the cloned Tranquillyzer repository.

## Usage

```bash
tranquillyzer annotate-reads \
    [OPTIONS] \
    OUTPUT_DIR \
    WHITELIST_FILE
```

## Command Line Arguments

Required Arguments

- `OUTPUT_DIR`: Path to your OUTPUT directory
- `WHITELIST_FILE`: TSV file containing the sequences that define each sample (e.g., the cell barcode (CBC) and/or dual
  index sequences). **Note**, the column names for these sequences must match the name used for that element in the
  model training. For example, if the model uses CBC to denote the cell barcode, the `WHITELIST_FILE` must have a column
  named "CBC" (not "barcode" or "cell_barcode"), though the order of the columns within the file does not matter.

Optional Arguments

- `--output-fmt TEXT`: Output format for demultiplexed reads: `fasta` or `fastq` (default: `fasta`)
- `--model-name TEXT:` Base model name (the name of the model without any suffixes) (default: `10x3p_sc_ont_011`)
  For `--model-type CRF`, `_w_CRF` will be appended to the base model name
- `--model-type TEXT`: The type of model to run (see above for descriptions of each type): `REG`, `CRF`, or `HYB`
  (default: `HYB`)
- `--strand TEXT`: Strand information of the reads: `3'`, `5'`, or `unstranded` (default: `unstranded`)
- `--chunk-size INTEGER`: Base chunk size for processing, dynamically adjusts based on read length (default: 100000)
- `--bc-lv-threshold INTEGER`: Levenshtein distance threshold for barcode correction (default: 2)
- `--threads INTEGER`: Number of CPU threads for barcode correction and demultiplexing (default: 12)
- `--max-queue-size INTEGER`: Max number of Parquet files to queue for post-processing (default: 3)
- `--help`: Print help message and exit

# Alignment

## Overview

Tranquillyzer calls `minimap2` to align the demultiplexed reads. It outputs a coordinate-sorted BAM and associated BAM
index file in `OUTPUT_DIR/aligned_files/`.

## Usage

```bash
tranquillyer align \
    [OPTIONS] \
    INPUT_DIR \
    REFERENCE \
    OUTPUT_DIR
```

## Command Line Arguments

Required Arguments

- `INPUT_DIR`: Path to annotated read output. In practice, this is the same directory given as the `OUTPUT_DIR` in the
  `annotate-reads` call. Looks for a file called `INPUT_DIR/demuxed_fasta/demuxed.fasta`.
- `REFERENCE`: Reference FASTA used for `minimap2`
- `OUTPUT_DIR`: Path to write OUTPUT directory

Optional Arguments

- `--preset TEXT`: `minimap2` preset (i.e., `minimap2 -ax <preset> ...` (default: None)
- `--filt-flag INT`: Flag for filtering reads via `samtools view -F <INT> ...`. Default is to filter out secondary
  alignments and unmapped reads (default: 260)
- `--mapq INTEGER`: Minimum MAPQ for the alignments to be included for the downstream analysis (default: 0)
- `--threads INTEGER`: Number of CPU threads (default: 12)
- `--add-minimap-args TEXT`: additional `minimap2` arguments (`-t` and `-ax <preset>` already included)
- `--help`: Show this message and exit

# Duplicate Marking

TODO: Fill out this section

## Overview

## Usage

```bash
```

## Command Line Arguments

Required Arguments

-

Optional Arguments

-

# Read Visualization

TODO: Fill out this section

## Overview

## Usage

```bash
```

## Command Line Arguments

Required Arguments

-

Optional Arguments

-

# Simulate Training Data

TODO: Fill out this section

## Overview

## Usage

```bash
```

## Command Line Arguments

Required Arguments

-

Optional Arguments

-

# Model Training

Coming soon!
